File: apps/roi-studio/roi_studio/core/store.py
"""The annotation store: one canonical row schema and every output file.

Rows are kept in ONE shape (CANON_COLUMNS) and projected to the chosen view
only at write time, which is what stops a format toggle from producing a
ragged mixed-schema spreadsheet.

Deliberately free of Qt so the whole persistence path is testable headlessly.
"""

from __future__ import annotations

import csv
import json
import os
import shutil
import tempfile
from datetime import datetime

from ..config import (APP_VERSION, CANON_COLUMNS, COLUMN_WIDTHS, FULL_VIEW,
                      JSON_NAME, MAP_NAME, NDIGITS, ROW_TYPES, TEXT_COLUMNS,
                      XLSX_NAME)
from . import geometry as geo
from .io_safe import (WriteReport, atomic_write, timestamped_sibling,
                      write_text_atomic)

try:
    from openpyxl import Workbook, load_workbook
    from openpyxl.styles import Alignment, Font, PatternFill
    from openpyxl.utils import get_column_letter
    HAS_XLSX = True
except Exception:                                    # pragma: no cover
    Workbook = load_workbook = None
    Font = PatternFill = Alignment = get_column_letter = None
    HAS_XLSX = False

LEGACY_CSV_NAMES = ("roi_annotations.csv",)

SPREADSHEET_EXTS = (".xlsx", ".xlsm", ".xltx", ".xltm")


def load_workbook_rows(path):
    """Read a workbook as a list of row tuples.

    openpyxl validates the *file name*, not the content, so it refuses a
    rolling backup called `roi_annotations.xlsx.bak` even though the bytes
    are a perfectly good workbook.  Anything without a supported extension is
    copied to a correctly named temporary file first."""
    if not HAS_XLSX:
        return []
    path = str(path)
    temporary = ""
    try:
        if not path.lower().endswith(SPREADSHEET_EXTS):
            fd, temporary = tempfile.mkstemp(suffix=".xlsx")
            os.close(fd)
            shutil.copyfile(path, temporary)
            path = temporary
        wb = load_workbook(path, read_only=True, data_only=True)
        try:
            return list(wb.active.iter_rows(values_only=True))
        finally:
            wb.close()
    finally:
        if temporary:
            try:
                os.remove(temporary)
            except OSError:
                pass


def rows_to_dicts(rows):
    """Turn raw workbook rows into dicts keyed by the header line."""
    if not rows:
        return []
    header = [str(h).strip() if h is not None else "" for h in rows[0]]
    out = []
    for values in rows[1:]:
        if values is None or all(v is None or str(v).strip() == ""
                                 for v in values):
            continue
        item = {}
        for i, key in enumerate(header):
            if not key:
                continue
            item[key] = "" if i >= len(values) or values[i] is None \
                else str(values[i])
        out.append(item)
    return out


# ══════════════════════════════════════════════════════════════
# JSON LAYOUT HELPERS
# ══════════════════════════════════════════════════════════════
def _compact_map_block(mapping, item_indent="  ", close_indent=""):
    """Render {"KEY": [[[x, y], ...]], ...} with ONE LINE PER KEY.

    json.dump(indent=2) puts every single number on its own line, which makes
    a 500-camera map unreadable.  Keys are still json.dumps()-escaped, so this
    stays valid JSON, and every file is parsed back before it replaces the
    old one anyway."""
    if not mapping:
        return "{}"
    parts = ["%s%s: %s" % (item_indent, json.dumps(k),
                           json.dumps(v, separators=(", ", ", ")))
             for k, v in mapping.items()]
    return "{\n" + ",\n".join(parts) + "\n" + close_indent + "}"


def dump_map_json(roi_map) -> str:
    """The map file: one roi_key per line."""
    return _compact_map_block(roi_map, "    ", "") + "\n"


def dump_full_json(full) -> str:
    """Metadata pretty-printed, coordinate maps compact."""
    lines = ["{"]
    skip = ("rois", "no_roi", "sources", "shapes")
    for key in [k for k in full if k not in skip]:
        lines.append('  %s: %s,' % (json.dumps(key), json.dumps(full[key])))
    lines.append('  "rois": %s,' % _compact_map_block(full["rois"], "    ", "  "))
    lines.append('  "shapes": %s,' % _compact_map_block(full["shapes"], "    ", "  "))
    lines.append('  "no_roi": %s,'
                 % json.dumps(full["no_roi"], separators=(", ", ", ")))
    lines.append('  "sources": %s' % _compact_map_block(full["sources"],
                                                        "    ", "  "))
    lines.append("}")
    return "\n".join(lines) + "\n"


# ══════════════════════════════════════════════════════════════
class AnnotationStore:
    """Owns every row and every output file for one batch folder."""

    def __init__(self):
        self.folder = ""
        self.rows = []
        self.xlsx_path = ""
        self.json_path = ""
        self.map_path = ""
        self.stamp = None            # mtime of the spreadsheet we last saw

    # ── setup ─────────────────────────────────────────────
    def bind(self, folder) -> None:
        self.folder = str(folder)
        self.xlsx_path = os.path.join(self.folder, XLSX_NAME)
        self.json_path = os.path.join(self.folder, JSON_NAME)
        self.map_path = os.path.join(self.folder, MAP_NAME)
        self.rows = []

    # ── canonical row construction ────────────────────────
    @staticmethod
    def make_row(fname, row_type, polys_px, norm_polys, comment,
                 width=0, height=0, kinds=None):
        site_id, cam_token = geo.extract_site_cam(fname)
        return {
            "image_name": fname,
            "roi_key": geo.make_roi_key(fname),
            "site_id": site_id,
            "cam_number": cam_token,
            "pixel_coords": geo.fmt_polys(polys_px),
            "normalized_coords": geo.fmt_polys(norm_polys),
            "total_polygons": len(polys_px) if row_type == "roi" else 0,
            "image_width": int(width or 0),
            "image_height": int(height or 0),
            "row_type": row_type if row_type in ROW_TYPES else "roi",
            "comment": comment or "",
            "shape_types": geo.fmt_shape_types(kinds) if row_type == "roi" else "",
        }

    @staticmethod
    def canonicalize(raw):
        """Bring any legacy row up to the current schema.

        Older builds wrote either the site layout (roi_coordinate) or the full
        layout (pixel_coords / normalized_coords), and toggling between them
        left rows carrying both.  Everything folds back into one shape here."""
        row = {k: "" for k in CANON_COLUMNS}
        for key in CANON_COLUMNS:
            if key in raw and raw[key] is not None:
                row[key] = str(raw[key]).strip()

        fname = row["image_name"]
        if not fname:
            # some very old files put the bare basename in site_id
            fname = str(raw.get("site_id", "") or "").strip()
            row["image_name"] = fname

        if not row["normalized_coords"]:
            legacy = raw.get("roi_coordinate", "")
            if legacy:
                polys = geo.parse_multi_polys(legacy)
                if polys:
                    row["normalized_coords"] = geo.fmt_polys(polys)

        # re-render both coordinate cells in the current bracket format
        for key in ("pixel_coords", "normalized_coords"):
            if row[key]:
                row[key] = geo.fmt_polys(geo.parse_multi_polys(row[key]))

        if fname:
            site_id, cam_token = geo.extract_site_cam(fname)
            row["site_id"] = row["site_id"] or site_id
            row["cam_number"] = row["cam_number"] or cam_token
            row["roi_key"] = row["roi_key"] or geo.make_roi_key(fname)

        rtype = row["row_type"] or "roi"
        row["row_type"] = rtype if rtype in ROW_TYPES else "roi"

        count = len(geo.parse_multi_polys(row["pixel_coords"]
                                          or row["normalized_coords"]))
        try:
            row["total_polygons"] = int(float(row["total_polygons"] or 0))
        except (TypeError, ValueError):
            row["total_polygons"] = count
        for key in ("image_width", "image_height"):
            try:
                row[key] = int(float(row[key] or 0))
            except (TypeError, ValueError):
                row[key] = 0

        # shape_types is padded to the polygon count, so a file written
        # before the column existed describes plain polygons
        row["shape_types"] = geo.fmt_shape_types(
            geo.parse_shape_types(row["shape_types"], count)) if count else ""

        # The JSON export is built from normalized_coords.  A legacy row that
        # only has pixels would otherwise vanish from the JSON, so derive them
        # whenever the image size is recoverable.
        if not row["normalized_coords"] and row["pixel_coords"] \
                and row["image_width"] and row["image_height"]:
            px = geo.parse_multi_polys(row["pixel_coords"])
            if px:
                row["normalized_coords"] = geo.fmt_polys(
                    geo.polys_to_norm(px, row["image_width"], row["image_height"]))
        return row

    # ── loading ───────────────────────────────────────────
    def load(self):
        """Prefer the xlsx, fall back to its .bak, then to a legacy csv.

        Returns a list of human-readable notes."""
        notes = []
        candidates = [(self._load_xlsx, self.xlsx_path, XLSX_NAME),
                      (self._load_xlsx, self.xlsx_path + ".bak", XLSX_NAME + ".bak")]
        for name in LEGACY_CSV_NAMES:
            path = os.path.join(self.folder, name)
            candidates.append((self._load_csv, path, name))
            candidates.append((self._load_csv, path + ".bak", name + ".bak"))

        for loader, path, label in candidates:
            if not path or not os.path.isfile(path):
                continue
            try:
                raw_rows = loader(path)
            except Exception as exc:
                notes.append("Could not read %s (%s)" % (label, exc))
                continue
            if raw_rows:
                self.rows = [self.canonicalize(r) for r in raw_rows]
                self.rows = [r for r in self.rows if r["image_name"]]
                notes.append("Loaded %d existing row(s) from %s"
                             % (len(self.rows), label))
                self.touch_stamp()
                return notes
        self.touch_stamp()
        return notes

    # ── external change detection ─────────────────────────
    def touch_stamp(self) -> None:
        """Remember the spreadsheet as it is on disk right now."""
        try:
            self.stamp = os.path.getmtime(self.xlsx_path) \
                if os.path.isfile(self.xlsx_path) else None
        except OSError:
            self.stamp = None

    def changed_externally(self) -> bool:
        """True when the spreadsheet on disk is not the one we last wrote.

        Someone else - another session, a sync client, a person editing in
        Excel - has been here since, so overwriting it would discard their
        work silently."""
        if not self.xlsx_path:
            return False
        try:
            exists = os.path.isfile(self.xlsx_path)
        except OSError:
            return False
        if not exists:
            return False
        if self.stamp is None:
            return True
        try:
            return abs(os.path.getmtime(self.xlsx_path) - self.stamp) > 1.0
        except OSError:
            return False

    @staticmethod
    def _load_csv(path):
        with open(path, "r", encoding="utf-8-sig", newline="") as fh:
            return [{k: ("" if v is None else v) for k, v in row.items()
                     if k is not None} for row in csv.DictReader(fh)]

    @staticmethod
    def _load_xlsx(path):
        return rows_to_dicts(load_workbook_rows(path))

    # ── row helpers ───────────────────────────────────────
    def purge(self, fname) -> None:
        self.rows = [r for r in self.rows if r["image_name"] != fname]

    def add(self, row) -> None:
        self.rows.append(row)

    def rows_for(self, fname):
        return [r for r in self.rows if r["image_name"] == fname]

    def row_for(self, fname):
        rows = self.rows_for(fname)
        return rows[0] if rows else None

    def image_names(self):
        seen, out = set(), []
        for row in self.rows:
            name = row["image_name"]
            if name not in seen:
                seen.add(name)
                out.append(name)
        return out

    # ── projection ────────────────────────────────────────
    @staticmethod
    def project(row, view):
        out = {}
        for key in view:
            if key == "roi_coordinate":
                out[key] = row.get("normalized_coords", "")
            else:
                out[key] = row.get(key, "")
        return out

    # ── JSON build ────────────────────────────────────────
    def build_json(self, rows=None):
        """Merge every image's polygons under its roi_key.

        Two images that share a site+camera (different timestamps) have their
        polygon lists concatenated, in natural filename order.  A key that has
        at least one ROI never appears in the no_roi list."""
        rows = self.rows if rows is None else rows
        roi_map, shape_map, sources = {}, {}, {}
        no_roi_keys, warnings = [], []

        ordered = sorted(rows, key=lambda r: geo.natural_key(r["image_name"]))
        for row in ordered:
            key = str(row.get("roi_key", "")).strip()
            fname = row.get("image_name", "")
            if not key:
                warnings.append("no roi_key for %s" % fname)
                continue
            rtype = row.get("row_type", "")
            if rtype == "roi":
                polys = geo.parse_multi_polys(row.get("normalized_coords", ""))
                if not polys:
                    # An ROI row that cannot be normalised would vanish from
                    # the JSON.  Say so rather than quietly shipping a short
                    # file - re-saving the image repairs it.
                    if row.get("pixel_coords"):
                        warnings.append(
                            "%s has pixel coords but no image size - "
                            "re-save it to include it in the JSON" % fname)
                    continue
                if key in roi_map:
                    warnings.append("merged %s into %s" % (fname, key))
                roi_map.setdefault(key, []).extend(polys)
                shape_map.setdefault(key, []).extend(
                    geo.parse_shape_types(row.get("shape_types", ""), len(polys)))
                sources.setdefault(key, []).append(fname)
            elif rtype == "no_roi":
                if key not in no_roi_keys:
                    no_roi_keys.append(key)
                sources.setdefault(key, []).append(fname)

        no_roi_keys = [k for k in no_roi_keys if k not in roi_map]

        rounded = {}
        for key in sorted(roi_map, key=geo.natural_key):
            rounded[key] = [[[round(float(x), NDIGITS), round(float(y), NDIGITS)]
                             for x, y in poly] for poly in roi_map[key]]

        full = {
            "generated_at": datetime.now().isoformat(timespec="seconds"),
            "generator": "ROI Studio v%s" % APP_VERSION,
            "source_folder": self.folder,
            "coordinate_space": "normalized",
            "decimals": NDIGITS,
            "roi_count": len(rounded),
            "no_roi_count": len(no_roi_keys),
            "total_polygons": sum(len(v) for v in rounded.values()),
            "rois": rounded,
            "shapes": {k: shape_map.get(k, []) for k in sorted(rounded, key=geo.natural_key)},
            "no_roi": sorted(no_roi_keys, key=geo.natural_key),
            "sources": {k: sources.get(k, []) for k in sorted(sources, key=geo.natural_key)},
        }
        return full, rounded, warnings

    # ── writing ───────────────────────────────────────────
    def flush(self, view_columns=None, rows=None) -> WriteReport:
        """Write the xlsx, the full json and the map json.

        Every file goes through atomic_write and is read back to prove it
        parses before the old copy is replaced."""
        report = WriteReport()
        if not self.folder:
            report.errors.append("no output folder")
            return report
        if not os.path.isdir(self.folder):
            report.errors.append("output folder no longer exists")
            return report

        rows = self.rows if rows is None else rows
        view_columns = list(view_columns or FULL_VIEW)

        full, roi_map, warns = self.build_json(rows)
        report.warnings.extend(warns[:3])

        self._write_json(self.json_path, dump_full_json(full), report)
        self._write_json(self.map_path, dump_map_json(roi_map), report)

        if HAS_XLSX:
            self._write_xlsx(view_columns, rows, report)
        else:
            report.errors.append(
                "openpyxl is not installed, so no spreadsheet can be written")
        return report

    def _write_json(self, path, text, report) -> None:
        ok, err = write_text_atomic(path, text, verify_json=True)
        if ok:
            report.written.append(path)
        else:
            report.errors.append("%s: %s" % (os.path.basename(path), err))

    def _write_xlsx(self, view_columns, rows, report) -> None:
        projected = [self.project(r, view_columns) for r in rows]

        def write(tmp):
            wb = Workbook()
            ws = wb.active
            ws.title = "ROI"
            ws.append(list(view_columns))
            for row in projected:
                ws.append([row.get(c, "") for c in view_columns])

            head_font = Font(bold=True, color="FFFFFF")
            head_fill = PatternFill("solid", fgColor="2F6BD8")
            for idx, name in enumerate(view_columns, start=1):
                cell = ws.cell(row=1, column=idx)
                cell.font = head_font
                cell.fill = head_fill
                cell.alignment = Alignment(horizontal="left", vertical="center")
                ws.column_dimensions[get_column_letter(idx)].width = \
                    COLUMN_WIDTHS.get(name, 16)
                # Force TEXT on every string column so Excel can never
                # re-interpret a coordinate string as numbers or dates.
                if name in TEXT_COLUMNS:
                    for r in range(2, len(projected) + 2):
                        ws.cell(row=r, column=idx).number_format = "@"
            ws.freeze_panes = "A2"
            if projected:
                ws.auto_filter.ref = "A1:%s%d" % (
                    get_column_letter(len(view_columns)), len(projected) + 1)
            wb.save(tmp)

        def verify(tmp):
            wb = load_workbook(tmp, read_only=True)
            try:
                if wb.active.max_row < 1:
                    raise ValueError("workbook came back empty")
            finally:
                wb.close()

        ok, err = atomic_write(self.xlsx_path, write, verify)
        if ok:
            report.written.append(self.xlsx_path)
            self.touch_stamp()
            return
        # The usual cause is the file being open in Excel; keep the work by
        # writing a timestamped sibling rather than losing the save.
        fallback = timestamped_sibling(self.xlsx_path)
        ok2, err2 = atomic_write(fallback, write, verify, keep_backup=False)
        if ok2:
            self.xlsx_path = fallback
            report.warnings.append("%s locked, wrote %s"
                                   % (XLSX_NAME, os.path.basename(fallback)))
            report.written.append(fallback)
            self.touch_stamp()
        else:
            report.errors.append("xlsx: %s / %s" % (err, err2))

File: packages/roi-core/src/store.ts
/**
 * The canonical row and the output files built from it.
 *
 * Port of the pure half of `apps/roi-studio/roi_studio/core/store.py`. The
 * desktop app's `AnnotationStore` also owns atomic writes, a folder lock and
 * external-change detection; none of that is meaningful in a browser, so this
 * module stops at producing the file *text* and leaves delivery to the caller.
 *
 * One column is new here: `shape_classes`. The desktop app has no notion of
 * classes, so it is written only when a batch actually uses one, keeping output
 * byte-identical to the desktop app's for un-classed batches.
 */

import {
  APP_VERSION,
  CANON_COLUMNS,
  DEFAULT_CLASS,
  NDIGITS,
  ROW_TYPES,
  type RowType,
} from "./config.ts";
import * as geo from "./geometry.ts";
import type { MultiPoly } from "./geometry.ts";

export interface CanonRow {
  image_name: string;
  roi_key: string;
  site_id: string;
  cam_number: string;
  pixel_coords: string;
  normalized_coords: string;
  total_polygons: number;
  image_width: number;
  image_height: number;
  row_type: RowType;
  comment: string;
  shape_types: string;
  shape_classes: string;
}

export interface MakeRowInput {
  fname: string;
  rowType: RowType;
  polysPx?: MultiPoly;
  normPolys?: MultiPoly;
  comment?: string;
  width?: number;
  height?: number;
  kinds?: readonly string[];
  classes?: readonly string[];
}

/** True when every class is the default, i.e. the batch does not use classes. */
function classesAreDefault(classes: readonly string[]): boolean {
  return classes.every((c) => !c || c === DEFAULT_CLASS);
}

export function makeRow(input: MakeRowInput): CanonRow {
  const {
    fname,
    rowType,
    polysPx = [],
    normPolys = [],
    comment = "",
    width = 0,
    height = 0,
    kinds = [],
    classes = [],
  } = input;
  const [siteId, camToken] = geo.extractSiteCam(fname);
  const isRoi = rowType === "roi";
  return {
    image_name: fname,
    roi_key: geo.makeRoiKey(fname),
    site_id: siteId,
    cam_number: camToken,
    pixel_coords: geo.fmtPolys(polysPx),
    normalized_coords: geo.fmtPolys(normPolys),
    total_polygons: isRoi ? polysPx.length : 0,
    image_width: Math.trunc(width || 0),
    image_height: Math.trunc(height || 0),
    row_type: (ROW_TYPES as readonly string[]).includes(rowType) ? rowType : "roi",
    comment: comment || "",
    shape_types: isRoi ? geo.fmtShapeTypes(kinds) : "",
    // only recorded when it carries information the desktop app would not
    shape_classes: isRoi && !classesAreDefault(classes) ? geo.jsonSpaced(classes) : "",
  };
}

/** Python's `int(float(x))`: parse as a number, then truncate towards zero. */
function intOfFloat(value: unknown, fallback: number): number {
  const text = String(value ?? "").trim();
  if (!text) return 0; // Python's `x or 0` makes a blank cell a zero, not the fallback
  const parsed = Number(text);
  if (!Number.isFinite(parsed)) return fallback;
  return Math.trunc(parsed);
}

/**
 * Bring any legacy or imported row up to the current schema.
 *
 * Older builds wrote either the site layout (`roi_coordinate`) or the full
 * layout, and toggling between them left rows carrying both. Everything folds
 * back into one shape here.
 */
export function canonicalize(raw: Record<string, unknown>): CanonRow {
  const text: Record<string, string> = {};
  for (const key of CANON_COLUMNS) {
    const value = raw[key];
    text[key] = value === null || value === undefined ? "" : String(value).trim();
  }

  let fname = text.image_name ?? "";
  if (!fname) {
    // some very old files put the bare basename in site_id
    fname = String(raw.site_id ?? "").trim();
  }

  let normalized = text.normalized_coords ?? "";
  if (!normalized) {
    const legacy = raw.roi_coordinate;
    if (legacy) {
      const polys = geo.parseMultiPolys(legacy);
      if (polys.length) normalized = geo.fmtPolys(polys);
    }
  }

  // re-render both coordinate cells in the current bracket format
  let pixel = text.pixel_coords ?? "";
  if (pixel) pixel = geo.fmtPolys(geo.parseMultiPolys(pixel));
  if (normalized) normalized = geo.fmtPolys(geo.parseMultiPolys(normalized));

  let siteId = text.site_id ?? "";
  let camNumber = text.cam_number ?? "";
  let roiKey = text.roi_key ?? "";
  if (fname) {
    const [derivedSite, derivedCam] = geo.extractSiteCam(fname);
    siteId = siteId || derivedSite;
    camNumber = camNumber || derivedCam;
    roiKey = roiKey || geo.makeRoiKey(fname);
  }

  const rawType = text.row_type || "roi";
  const rowType: RowType = (ROW_TYPES as readonly string[]).includes(rawType)
    ? (rawType as RowType)
    : "roi";

  const count = geo.parseMultiPolys(pixel || normalized).length;
  const width = intOfFloat(text.image_width, 0);
  const height = intOfFloat(text.image_height, 0);

  // The JSON export is built from normalized_coords. A legacy row that only
  // has pixels would otherwise vanish, so derive them when the size is known.
  if (!normalized && pixel && width && height) {
    const px = geo.parseMultiPolys(pixel);
    if (px.length) normalized = geo.fmtPolys(geo.polysToNorm(px, width, height));
  }

  const classes = parseShapeClasses(text.shape_classes, count);

  return {
    image_name: fname,
    roi_key: roiKey,
    site_id: siteId,
    cam_number: camNumber,
    pixel_coords: pixel,
    normalized_coords: normalized,
    total_polygons: intOfFloat(text.total_polygons, count),
    image_width: width,
    image_height: height,
    row_type: rowType,
    comment: text.comment ?? "",
    // padded to the polygon count, so a file written before the column
    // existed describes plain polygons
    shape_types: count ? geo.fmtShapeTypes(geo.parseShapeTypes(text.shape_types, count)) : "",
    shape_classes: count && !classesAreDefault(classes) ? geo.jsonSpaced(classes) : "",
  };
}

/**
 * Read the shape_classes cell, padded to `count` with the default class. A
 * missing cell is the normal case for anything the desktop app wrote.
 */
export function parseShapeClasses(cell: unknown, count = 0): string[] {
  const classes: string[] = [];
  if (cell) {
    let s = String(cell).trim();
    if (s.startsWith("'")) s = s.slice(1).trim();
    try {
      const parsed: unknown = JSON.parse(s);
      if (Array.isArray(parsed)) {
        for (const c of parsed) classes.push(String(c) || DEFAULT_CLASS);
      }
    } catch {
      // an unreadable cell simply means no classes were recorded
    }
  }
  if (count) {
    while (classes.length < count) classes.push(DEFAULT_CLASS);
    return classes.slice(0, count);
  }
  return classes;
}

// ── row collections ───────────────────────────────────────────

export function purgeRows(rows: readonly CanonRow[], fname: string): CanonRow[] {
  return rows.filter((r) => r.image_name !== fname);
}

export function rowsFor(rows: readonly CanonRow[], fname: string): CanonRow[] {
  return rows.filter((r) => r.image_name === fname);
}

export function rowFor(rows: readonly CanonRow[], fname: string): CanonRow | null {
  return rows.find((r) => r.image_name === fname) ?? null;
}

/** Unique image names, in first-seen order. */
export function imageNames(rows: readonly CanonRow[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const row of rows) {
    if (!seen.has(row.image_name)) {
      seen.add(row.image_name);
      out.push(row.image_name);
    }
  }
  return out;
}

/** Reduce a row to one spreadsheet layout's columns. */
export function project(
  row: CanonRow,
  view: readonly string[],
): Record<string, string | number> {
  const out: Record<string, string | number> = {};
  for (const key of view) {
    if (key === "roi_coordinate") {
      out[key] = row.normalized_coords ?? "";
    } else {
      out[key] = (row as unknown as Record<string, string | number>)[key] ?? "";
    }
  }
  return out;
}

// ── JSON build ────────────────────────────────────────────────

export interface FullJson {
  generated_at: string;
  generator: string;
  source_folder: string;
  coordinate_space: string;
  decimals: number;
  roi_count: number;
  no_roi_count: number;
  total_polygons: number;
  rois: Record<string, MultiPoly>;
  shapes: Record<string, string[]>;
  classes?: Record<string, string[]>;
  no_roi: string[];
  sources: Record<string, string[]>;
}

export interface BuildJsonResult {
  full: FullJson;
  map: Record<string, MultiPoly>;
  warnings: string[];
}

/** `datetime.now().isoformat(timespec="seconds")` — local time, no zone. */
function localIsoSeconds(date: Date): string {
  const pad = (n: number): string => String(n).padStart(2, "0");
  return (
    `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}` +
    `T${pad(date.getHours())}:${pad(date.getMinutes())}:${pad(date.getSeconds())}`
  );
}

/**
 * Merge every image's polygons under its roi_key.
 *
 * Two images that share a site and camera (different timestamps) have their
 * polygon lists concatenated, in natural filename order. A key with at least
 * one ROI never appears in the no_roi list.
 */
export function buildJson(
  rows: readonly CanonRow[],
  folder = "",
  now: Date = new Date(),
): BuildJsonResult {
  const roiMap = new Map<string, MultiPoly>();
  const shapeMap = new Map<string, string[]>();
  const classMap = new Map<string, string[]>();
  const sources = new Map<string, string[]>();
  const noRoiKeys: string[] = [];
  const warnings: string[] = [];

  const ordered = [...rows].sort((a, b) => geo.naturalCompare(a.image_name, b.image_name));
  for (const row of ordered) {
    const key = String(row.roi_key ?? "").trim();
    const fname = row.image_name ?? "";
    if (!key) {
      warnings.push(`no roi_key for ${fname}`);
      continue;
    }
    if (row.row_type === "roi") {
      const polys = geo.parseMultiPolys(row.normalized_coords);
      if (polys.length === 0) {
        // An ROI row that cannot be normalised would vanish from the JSON.
        // Say so rather than quietly shipping a short file.
        if (row.pixel_coords) {
          warnings.push(
            `${fname} has pixel coords but no image size - re-save it to include it in the JSON`,
          );
        }
        continue;
      }
      if (roiMap.has(key)) warnings.push(`merged ${fname} into ${key}`);
      roiMap.set(key, [...(roiMap.get(key) ?? []), ...polys]);
      shapeMap.set(key, [
        ...(shapeMap.get(key) ?? []),
        ...geo.parseShapeTypes(row.shape_types, polys.length),
      ]);
      classMap.set(key, [
        ...(classMap.get(key) ?? []),
        ...parseShapeClasses(row.shape_classes, polys.length),
      ]);
      sources.set(key, [...(sources.get(key) ?? []), fname]);
    } else if (row.row_type === "no_roi") {
      if (!noRoiKeys.includes(key)) noRoiKeys.push(key);
      sources.set(key, [...(sources.get(key) ?? []), fname]);
    }
    // a "comment" row contributes nothing to the JSON
  }

  const liveNoRoi = noRoiKeys.filter((k) => !roiMap.has(k));
  const roiKeysSorted = [...roiMap.keys()].sort(geo.naturalCompare);

  const rounded: Record<string, MultiPoly> = {};
  for (const key of roiKeysSorted) {
    rounded[key] = (roiMap.get(key) ?? []).map((poly) =>
      poly.map(([x, y]): [number, number] => [
        geo.pyRound(x, NDIGITS),
        geo.pyRound(y, NDIGITS),
      ]),
    );
  }

  const shapes: Record<string, string[]> = {};
  for (const key of roiKeysSorted) shapes[key] = shapeMap.get(key) ?? [];

  const sourcesOut: Record<string, string[]> = {};
  for (const key of [...sources.keys()].sort(geo.naturalCompare)) {
    sourcesOut[key] = sources.get(key) ?? [];
  }

  const full: FullJson = {
    generated_at: localIsoSeconds(now),
    generator: `ROI Studio v${APP_VERSION}`,
    source_folder: folder,
    coordinate_space: "normalized",
    decimals: NDIGITS,
    roi_count: roiKeysSorted.length,
    no_roi_count: liveNoRoi.length,
    total_polygons: roiKeysSorted.reduce((sum, k) => sum + (rounded[k]?.length ?? 0), 0),
    rois: rounded,
    shapes,
    no_roi: [...liveNoRoi].sort(geo.naturalCompare),
    sources: sourcesOut,
  };

  // Only carried when the batch actually uses classes, so an un-classed batch
  // produces exactly the file the desktop app would.
  const usesClasses = roiKeysSorted.some((k) =>
    (classMap.get(k) ?? []).some((c) => c && c !== DEFAULT_CLASS),
  );
  if (usesClasses) {
    const classes: Record<string, string[]> = {};
    for (const key of roiKeysSorted) classes[key] = classMap.get(key) ?? [];
    full.classes = classes;
  }

  return { full, map: rounded, warnings };
}

// ── serialisation ─────────────────────────────────────────────

/**
 * Coordinates in the JSON files are Python floats, so a whole number is
 * written `0.0` rather than `0`. (The spreadsheet cell is the opposite: it
 * keeps whole numbers whole, which is what `fmtPolys` does.) JavaScript has
 * only one number type, so the distinction has to be restored here.
 */
function jsonSpacedFloats(value: unknown): string {
  if (Array.isArray(value)) {
    return `[${value.map((item) => jsonSpacedFloats(item)).join(", ")}]`;
  }
  if (typeof value === "number") {
    return Number.isInteger(value) ? `${value}.0` : String(value);
  }
  return JSON.stringify(value);
}

/**
 * Render `{"KEY": [[[x, y], ...]], ...}` with one line per key.
 *
 * `JSON.stringify(value, null, 2)` puts every single number on its own line,
 * which makes a 500-camera map unreadable.
 */
function compactMapBlock(
  mapping: Record<string, unknown>,
  itemIndent = "  ",
  closeIndent = "",
  floats = false,
): string {
  const keys = Object.keys(mapping);
  if (keys.length === 0) return "{}";
  const render = floats ? jsonSpacedFloats : geo.jsonSpaced;
  const parts = keys.map((k) => `${itemIndent}${JSON.stringify(k)}: ${render(mapping[k])}`);
  return `{\n${parts.join(",\n")}\n${closeIndent}}`;
}

/** The map file: one roi_key per line. */
export function dumpMapJson(roiMap: Record<string, MultiPoly>): string {
  return `${compactMapBlock(roiMap, "    ", "", true)}\n`;
}

/** The full record: metadata pretty-printed, coordinate maps compact. */
export function dumpFullJson(full: FullJson): string {
  const skip = new Set(["rois", "no_roi", "sources", "shapes", "classes"]);
  const lines = ["{"];
  for (const key of Object.keys(full)) {
    if (skip.has(key)) continue;
    const value = (full as unknown as Record<string, unknown>)[key];
    lines.push(`  ${JSON.stringify(key)}: ${JSON.stringify(value)},`);
  }
  lines.push(`  "rois": ${compactMapBlock(full.rois, "    ", "  ", true)},`);
  lines.push(`  "shapes": ${compactMapBlock(full.shapes, "    ", "  ")},`);
  if (full.classes) {
    lines.push(`  "classes": ${compactMapBlock(full.classes, "    ", "  ")},`);
  }
  lines.push(`  "no_roi": ${geo.jsonSpaced(full.no_roi)},`);
  lines.push(`  "sources": ${compactMapBlock(full.sources, "    ", "  ")}`);
  lines.push("}");
  return `${lines.join("\n")}\n`;
}
